fuzzers/main.py

```diff
@@ -0,0 +1,642 @@
+        minimizer = f"cd {self.target['crate']}\nRUSTC_BOOTSTRAP=1 cargo fuzz tmin {self.target['runner']} artifacts/{self.target['runner']}/{artifact}"
+
+        # Check the artifact was not reported yet for this branch
+        global REPORTED_ARTIFACTS
+        if artifact in REPORTED_ARTIFACTS[branch]:
+            return
+        already_reported_for = []
+        for other_branch in REPORTED_ARTIFACTS.keys():
+            if artifact in REPORTED_ARTIFACTS[other_branch]:
+                already_reported_for.append(other_branch)
+        REPORTED_ARTIFACTS[branch].append(artifact)
+
+        if len(already_reported_for) == 0:
+            already_reported_msg = ""
+        else:
+            already_reported_msg = f"(already reported for branches {already_reported_for})"
+
+        # Send the information in two mesages, a short one guaranteed to succeed, then a long one
+        # with the log lines that could go over the message size limit
+        import textwrap
+        import zulip
+        client = zulip.Client(config_file="~/.fuzzer-zuliprc")
+        client.send_message({
+            "type": "stream",
+            "to": "nearinc/fuzzer/private",
+            "topic": f"{branch}: artifact {artifact}",
+            "content": (
+                f"# Fuzzer found new crash for branch *{branch}* {already_reported_msg}\n"
+                f"Full logs are available at {logs_url}.\n"
+                f"\n"
+                f"You can download the artifact by using the following command (all commands are to be run from the root of `nearcore`):\n"
+                f"```\n"
+                f"{downloader}\n"
+                f"```\n"
+                f"\n"
+                f"Then, you can reproduce by running the following command:\n"
+                f"```\n"
+                f"{reproducer}\n"
+                f"```\n"
+                f"\n"
+                f"Or minimize by running the following command:\n"
+                f"```\n"
+                f"{minimizer}\n"
+                f"```\n"
+                f"\n"
+                f"Please edit the topic name to add more meaningful information once investigated. Keeping the artifact hash in it can help if the same artifact gets detected as crashing another branch.\n"
+            ),
+        })
+        last_log_lines = ""
+        for l in log_lines[::-1]:
+            if l.startswith("```"):
+                # Censor the end of a spoiler block, not great but this is for human consumption
+                # anyway
+                l = " " + l
+            if len(last_log_lines) + len(l) > 9000:
+                # Zulip limit is 10k, let's keep some safety buffer here
+                break
+            last_log_lines = l + last_log_lines
+        client.send_message({
+            "type": "stream",
+            "to": "nearinc/fuzzer/private",
+            "topic": f"{branch}: artifact {artifact}",
+            "content": f"```spoiler Last few log lines\n{last_log_lines}\n```",
+        })
```

Comment:
Just set up Grafana alert when the counter increases, no? You can then do whatever you want with the notification including sending it to Zulip or to Pager Duty.

Reply:
That's what I had initially done, but it turns out that during testing I had one finding which had quite a few artifacts appearing at once; so having the metadata directly available from zulip (in particular being able to match the log lines, and have reminders for how to fetch and reproduce the artifact locally) would have been quite helpful, avoiding having to dig into logs to figure out which artifact is which log files and how one should recover the proper artifact. Having the last few log lines straight in the crash report will probably come in handy too, as GCS serves as bytestream and thus wouldn't even allow reading the logs in the browser.

Also, hopefully we won't find crashes often, so having the documentation alongside the crash notification itself would probably be helpful :) the only setup needed is to have a `~/.fuzzer-zuliprc` file that can be downloaded straight from the zulip bot web interface to get the credentials working, and the code itself is barely more than the equivalent documentation code that'd have to be written otherwise.

Reply:
The instructions can be written in alert template and you can export failure count by branch and target. I’m not entirely convinced this is the best way of dealing with those notifications, but I guess we can go with it.